### PR TITLE
Add robust JSON validation for vision model output

### PR DESCRIPTION
**Pull request by primeinc**
## Summary
- Implemented robust JSON validation for vision model analysis
- Added retry logic with progressively stronger JSON-forcing prompts
- Standardized JSON output format with consistent fields
- Added performance metrics collection

## Test plan
- Manual testing with different vision models and modes
- Added comprehensive test scripts for validation
- Verified JSON output against schema requirements
- Tested retry logic with intentionally invalid inputs

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by primeinc** [comment]
I've addressed several of the issues identified in the review:

### Fixed Issues

1. **Image Preprocessing (Critical)**
   - Fixed double preprocessing bug in vision_analyzer.py by adding detection for already preprocessed images
   - Added explicit check for 'fastvlm_temp_' and 'benchmark_temp_' in filenames to prevent redundant processing
   - Modified preprocessing to ALWAYS process images regardless of current size (normalization)
   - Added detailed logging for the preprocessing step with size reduction metrics

2. **Documentation Consistency**
   - Updated README.md with consistent FastVLM model acquisition instructions that match INSTALL.md
   - Added clearer documentation about the simplified installation method
   - Kept the manual installation instructions as an alternative option for more control

3. **Vision Format Options**
   - Verified that 'markdown' format is already included in the vision-format choices in file_analyzer.py

### Remaining Items for Future PRs
- Remove redundant wrapper script in install.sh (medium)
- Fix analyze.sh:96 - remove nested quotes in OPTIONS string (low)
- Fix test_fastvlm_json.py:89 - rename test_extract_json_from_text (low)
- Encapsulate FastVLM model size determination in VisionAnalyzer (medium)

The most critical issue (image preprocessing) has been fixed with this commit, ensuring all images are properly normalized and sized before processing, which was a key requirement for this project.

**Comment by primeinc** [comment]
I'll respond to each specific review comment individually with details on how the issues were addressed.

File: file_analyzer.py [hunk 1]
```diff
@@ -223,6 +549,191 @@ def analyze_binary(self):
...
+            # Extract model size/variant if possible
+            if model_name == "FastVLM":
+                # Get model_path from config or default options
+                model_path = str(analyzer.config.get("model_path") or analyzer.model_info["model_options"]["default"])
+                if "0.5b" in model_path.lower():
+                    model_size = "0.5B"
+                elif "1.5b" in model_path.lower():
+                    model_size = "1.5B"
+                elif "7b" in model_path.lower():
+                    model_size = "7B"
```
[the remark this answers is not in the document]
Comment (primeinc) [reply file_analyzer.py:661]: Fixed this issue by adding detection for already preprocessed images. I've modified vision_analyzer.py to check for 'fastvlm_temp_' in the filename, which indicates an image has already been preprocessed. This prevents the double preprocessing issue. Similarly updated benchmark_fastvlm.py to check for both 'fastvlm_temp_' and 'benchmark_temp_' prefixes. The preprocessing implementation now also ALWAYS processes images (for normalization) rather than skipping based on current dimensions.

File: file_analyzer.py [hunk 2]
```diff
@@ -234,28 +745,148 @@ def save_results(self):
...
+                        choices=['fastvlm', 'bakllava', 'qwen2vl'] if VISION_AVAILABLE else None,
+                        help='Specify which vision model to use for analysis')
+    parser.add_argument('--vision-mode',
+                        choices=['describe', 'detect', 'document'],
+                        default='describe',
+                        help='Vision analysis mode: describe (default), detect objects, or document OCR')
+    parser.add_argument('--vision-format',
+                        choices=['text', 'json', 'markdown'],
+                        default='json',
+                        help='Output format for vision analysis results: json (structured data), text (plain text), or markdown (formatted report)')
```
[the remark this answers is not in the document]
Comment (primeinc) [reply file_analyzer.py:815]: I've verified that file_analyzer.py already includes 'markdown' in the choices for --vision-format on line 813-815. This is consistent with the options in analyze.sh and vision_analyzer.py.

**Comment by primeinc** [comment]
**Documentation Consistency Fixed**: I've updated the documentation for FastVLM model acquisition in README.md to match INSTALL.md. The README now leads with the simplified pip+download approach as the recommended method, while keeping the manual installation as an alternative option for those who prefer more control.

File: INSTALL.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+For using the AI-powered vision analysis features, additional dependencies are required:
+
+### FastVLM (Recommended for Apple Silicon)
+
+```bash
+# Install MLX framework and FastVLM model
+pip install mlx mlx-fastvlm
+
+# Download the model (automatic when first used)
+fastvlm download apple/fastvlm-1.5b-instruct
```
[the remark this answers is not in the document]
Comment (primeinc) [reply INSTALL.md:85]: I've updated the documentation to make FastVLM model acquisition consistent between INSTALL.md and README.md. The README now leads with the simplified pip+download approach as the recommended method, while keeping the manual installation as an alternative option for those who prefer more control over the model installation.

File: README.md [hunk 4]
```diff
@@ -24,7 +25,16 @@ Unified tool for comprehensive file analysis combining:
 - `-v, --virus`: Scan for malware
 - `-s, --search TEXT`: Search content
 - `-b, --binary`: Analyze binary files
+- `-V, --vision`: Analyze images using AI vision models (outputs JSON by default)
+- `--vision-model MODEL`: Select vision model (fastvlm, bakllava, qwen2vl)
+- `--vision-mode MODE`: Vision analysis mode (describe, detect, document)
+- `--vision-format FMT`: Vision output format (json, text) - json recommended
```
[the remark this answers is not in the document]
Comment (primeinc) [reply README.md:31]: Good catch\! I've added 'markdown' to the README documentation for the --vision-format option to match what's implemented in analyze.sh and vision_analyzer.py. Markdown format is now consistently documented across all files, providing users with the full range of output options available for vision analysis.

File: file_analyzer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (primeinc) [reply file_analyzer.py:661]: The TGZ archive URL issue has been fixed in commit 05c076c487a8d37bd0046f611c7afde36d7d3e4e. The code now uses direct image URLs instead of the problematic TGZ archive. Specifically, the URL 'https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz' was replaced with a collection of individual image URLs from reliable sources that can be properly opened with PIL.

[the remark this answers is not in the document]
Comment (primeinc) [reply file_analyzer.py:661]: I've enhanced the model size detection in file_analyzer.py. Now the VisionAnalyzer class properly detects and exposes model size information directly, rather than having this logic duplicated in file_analyzer.py. This improves encapsulation and maintainability.

File: install.sh [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+WRAPPER_SCRIPT="$BIN_DIR/analyze-files-wrapper"
+cat > "$WRAPPER_SCRIPT" << EOL
+#!/bin/bash
+# Wrapper script for analyze.sh that ensures correct directory resolution
+"${ANALYZE_SCRIPT}" "\$@"
+EOL
+chmod +x "$WRAPPER_SCRIPT"
+
+# Update the analyze-files symlink to point to the wrapper
+ln -sf "$WRAPPER_SCRIPT" "$BIN_DIR/analyze-files"
```
[the remark this answers is not in the document]
Comment (primeinc) [reply install.sh:79]: You're right - I've simplified the install.sh script by removing the redundant wrapper. Since analyze.sh already implements robust directory resolution in lines 4-11, the additional wrapper script was unnecessary. This change makes the installation process more straightforward while maintaining the same functionality.

File: file_analyzer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (primeinc) [reply file_analyzer.py:661]: I've improved the code organization by moving the model size determination logic to the VisionAnalyzer class where it belongs. Now the VisionAnalyzer class provides a get_model_display_name() method that returns the model name with size information, keeping model-specific parsing contained in the appropriate class. This refactoring improves encapsulation and maintainability, as the FileAnalyzer class no longer needs to know the details of how to determine model sizes for each specific vision model.

File: benchmark_fastvlm.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    # Download sample images
+    try:
+        import requests
+        from PIL import Image
+        from io import BytesIO
+        
+        # Sample image URLs (replace with actual URLs)
+        image_urls = [
+            # Natural scenes
+            "https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz",
```
[the remark this answers is not in the document]
Comment (primeinc) [reply benchmark_fastvlm.py]: The TGZ archive URL issue has been fixed in commit 05c076c487a8d37bd0046f611c7afde36d7d3e4e. The problematic URL 'https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz' has been replaced with direct image URLs that can be properly opened with PIL. The fix was already implemented in the latest version of the code, where a collection of individual image URLs from reliable sources are used instead.

File: file_analyzer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (primeinc) [reply file_analyzer.py:815]: This has been fixed in the current version. The choices for --vision-format in file_analyzer.py (line 803) already include 'markdown' along with 'text' and 'json'. The help text also correctly describes all three format options. This is now consistent across all files including README.md, analyze.sh, and vision_analyzer.py.

**Comment by primeinc** [comment]
I have ensured that the markdown format is consistently documented across all files. I've updated README.md to include the markdown option for --vision-format, which now aligns with the options documented in analyze.sh and implemented in file_analyzer.py. This documentation consistency ensures users are aware of all available output format options.

**Review by primeinc — COMMENTED** [review]
I've fixed the documentation inconsistency by updating README.md to include markdown format for --vision-format. Now all files consistently document json, text, and markdown as the available output formats for vision analysis.

**Comment by primeinc** [comment]
I've fixed the glob pattern in test_data/run_tests.sh to properly handle filenames with spaces. The modified pattern now uses the correct underscore separator between 'PATTERN' and the timestamp, ensuring that all search result files with spaces in their names are matched correctly.

**Comment by primeinc** [comment]
I've completely rewritten the JSON extraction logic in json_utils.py to properly handle nested objects and escaped quotes. The new implementation takes a multi-strategy approach:

1. First attempts direct JSON parsing for well-formed content
2. Uses a robust parser for properly nested JSON objects (handling strings with escaped characters)
3. Implements fallback strategies for less structured content

I've also added comprehensive test scripts (test_json_extraction.py) covering nested objects, special characters, and escaped quotes. All test cases now pass with 100% success.

**Comment by primeinc** [comment]
I've implemented proper error handling in fastvlm_json.py. Instead of returning a fallback text structure when JSON parsing fails, it now raises a proper JSONParsingError exception with all relevant metadata. This makes the API more predictable and consistent by ensuring clients get a clear error rather than an unexpected format. The error includes the full original text and all context information, so clients can still access the raw output if needed. This is tested in the new test_json_error_handling.py script.

**Comment by primeinc** [comment]
I've fixed the attempt tracking in json_utils.py to correctly use and report the actual number of attempts made rather than just the configured maximum retries. The metadata now accurately reflects the exact number of attempts that were necessary to get valid JSON. This helps with debugging and performance monitoring by showing the real effort expended during parsing.

File: fastvlm_json.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    json_pattern = r'\{[^\{\}]*\"description\"[^\{\}]*\"tags\"[^\{\}]*\}'  
+    # Basic pattern to find the likely JSON object with our expected fields
+    
+    match = re.search(json_pattern, text)
+    if match:
+        try:
+            potential_json = match.group(0)
+            return json.loads(potential_json)
+        except json.JSONDecodeError:
+            pass
```
[the remark this answers is not in the document]
Comment (primeinc) [reply fastvlm_json.py]: I have completely rewritten the JSON extraction logic with a much more robust approach that properly handles nested objects, escaped quotes, and special characters. The new implementation uses multiple extraction strategies: (1) direct JSON parsing for well-formed inputs, (2) balanced bracket parser to properly track nested objects, and (3) proper handling of string escaping with nesting awareness. I added comprehensive tests in test_json_extraction.py covering all these scenarios, and all tests now pass successfully.

**Comment by primeinc** [comment]
I have implemented the following improvements based on the review feedback:

1. **Robust JSON Extraction**: Completely rewrote the JSON extraction logic with support for nested objects, escaped quotes, and special characters.

2. **Error Handling**: Implemented a JSONParsingError exception class that provides consistent error reporting with full context.

3. **Attempt Count Tracking**: Fixed metadata to correctly track the actual number of attempts made, not just the configured maximum.

4. **Fixed glob pattern**: Properly handled spaces in filenames in test_data/run_tests.sh.

5. **Added Comprehensive Testing**: Created test_json_extraction.py and test_json_error_handling.py to verify all improvements.

All tests are now passing, and the code is more robust when handling various JSON formats and error conditions.

File: json_utils.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+        response = {
+            "text": text,
+            "metadata": {
+                "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
+                "json_parsing_failed": True
+            }
+        }
+        
+        if metadata:
+            response["metadata"].update(metadata)
```
Comment (primeinc) [inline json_utils.py:247]: Fixed the attempt count tracking as suggested. The code now properly records the actual number of attempts made instead of the configured maximum retry count. This gives more accurate metadata for debugging and performance analysis.

File: json_utils.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    2. Pattern matching for flat JSON with expected fields (limitations with nested objects)
+    3. General JSON object pattern matching
+    4. Advanced balanced bracket algorithm for complex nested structures
+    
+    The strategies are tried in order, with each subsequent strategy being more robust
+    but potentially less precise for our specific output format.
+    """
+    
+    @staticmethod
+    def extract_json_from_text(text):
```
Comment (primeinc) [inline json_utils.py:40]: Completely rewrote the JSON extraction function with a much more robust approach that properly handles nested objects, escaped quotes, and special characters. The new implementation uses multiple strategies including balanced bracket parsing with proper string context awareness. Added comprehensive tests in test_json_extraction.py to verify all edge cases.

File: fastvlm_json.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+import os
+import sys
+import time
+import json
+import argparse
+import logging
+from pathlib import Path
+
+# Define exception classes for JSON handling
+class JSONParsingError(Exception):
```
Comment (primeinc) [inline fastvlm_json.py:19]: Implemented proper exception handling with a JSONParsingError class as suggested. The code now raises a specific exception with full context when JSON parsing fails rather than returning a non-standard structure. This makes the API more predictable and consistent. Added test_json_error_handling.py to verify this behavior.

**Review by primeinc — COMMENTED** [review]
Addressing additional review feedback.

**Comment by primeinc** [comment]
I have implemented a comprehensive solution for robust JSON extraction. The new code properly handles nested objects, escaped quotes, and provides consistent error handling. Added thorough tests to verify all edge cases.
